test/generator/configure_generator.cpp

```diff
@@ -0,0 +1,86 @@
+#include "Halide.h"
+
+namespace {
+
+class Configure : public Halide::Generator<Configure> {
+public:
+    GeneratorParam<int> num_extra_buffer_inputs{"num_extra_buffer_inputs", 3};
+
+    Input<Buffer<int>> input{"input", 3};
+    Input<int> bias{"bias"};
+
+    Output<Buffer<int>> output{ "output", 3 };
+
+    void configure() {
+        configure_calls++;
+
+        // It's fine to examine GeneratorParams in the configure() method.
+        assert(num_extra_buffer_inputs == 3);
+
+        // Pointers returned by add_input() are managed by the Generator;
+        // user code must not free them. We can stash them in member variables
+        // as-is or in containers, like so:
+        for (int i = 0; i < num_extra_buffer_inputs; ++i) {
+            auto *extra = add_input<Buffer<>>("extra_" + std::to_string(i), UInt(8), 2);
+            extra_buffer_inputs.push_back(extra);
+        }
+
+        typed_extra_buffer_input = add_input<Buffer<int16_t>>("typed_extra_buffer_input", 2);
```

Comment:
`add_input<T>` is fine, but what would be even better is an interface like `add_input(std::string descriptor)`.
This would for a true, flexible and generic meta-programming facility within `configure()`, without the having to rely on the horrendous C++ type system and template machinery.
(btw, we can still have `add_input<T>` as a wrapper around `add_input(std::string descriptor)`).

Reply:
What does `descriptor` look like?

Reply:
But is the input a Buffer, a Func, or some scalar? If you are going to use the input in the generate() method it's going to need to have a c++ type.

Reply:
And yes, we'd need a "generic" `Input<void>` type to allow for fully customizable types.

Reply:
There would have to be some piece of code that parses these string descriptions and dynamically creates objects with the correct actual C++ type to use in the generate() method. The idea of add_input is that it's the method that code or something similar would call. It's to enable metaprogramming of generators in a type-safe way. You can always layer something like generators-specified-by-json-strings on top of it, but the underlying fully typed layer would still be necessary.

Reply:
Ah, indeed, I guess I can just build these abstractions on top of the typed layer.